### PR TITLE
Cmake

### PR DESCRIPTION
**Pull request by rdaly525**
This PR is Cmakeifying CoreIR. 

Tests are *not* being done using cmake currently. 

A large chunk of the files changed are due to fixing warnings when building the release version of cmake. These are fixed by changing assert -> ASSERT and by using a new coreir_unreachable() macro.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    # require at least gcc 4.8
+    if (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 4.9)
+        message(FATAL_ERROR "GCC version must be at least 4.9!")
+    endif()
+else()
+    message(WARNING "You are using an untested compiler")
+endif()
+
+set(CMAKE_CXX_FLAGS "-std=c++11 -Wall -fPIC -Werror")
+#CXXFLAGS = -std=c++11 -Wall -fPIC -Werror -ferror-limit=5
```
Comment (leonardt) [inline CMakeLists.txt]: Do we need this comment?

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+add_custom_command(
+  TARGET coreir-simulator
+  POST_BUILD
+  COMMAND ${CMAKE_COMMAND} -E copy "$<TARGET_FILE:coreir-simulator>" "${CMAKE_SOURCE_DIR}/bin/"
+  COMMENT "Copying to bin/"
+)
+
+
+#TARGETS coreir coreirsim coreir-commonlib
```
Comment (leonardt) [inline src/CMakeLists.txt]: Do we need this comment?

File: src/libs/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        DESTINATION lib
+    )
+  add_custom_command(
+    TARGET ${COREIRLIB}
+    POST_BUILD
+    COMMAND ${CMAKE_COMMAND} -E copy "$<TARGET_FILE:${COREIRLIB}>" "${CMAKE_SOURCE_DIR}/lib/"
+    COMMENT "Copying to lib/"
+  )
+
+endforeach()
```
Comment (leonardt) [inline src/libs/CMakeLists.txt]: foreach is a nice feature

**Comment by leonardt** [comment]
I think it's okay to merge this before tests have been migrated, since this seems like it will make it easier to work with the build system.

**Comment by rdaly525** [comment]
@leonardt, it does seem to incrementally rebuild appropriately when I edit some .h files.

**Pull request merged**